File: core/src/de/project/ice/inventory/Inventory.java
package de.project.ice.inventory;

import com.badlogic.gdx.utils.Array;
import de.project.ice.IceGame;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class Inventory
{
    @NotNull
    public final Array<Item> items = new Array<Item>();
    @NotNull
    private final IceGame game;

    public Inventory(@NotNull IceGame game)
    {
        this.game = game;
    }

    public void addItem(String name)
    {
        removeItem(name);
        Item item = Items.get(name);
        item.inventory = this;
        items.add(item);
    }

    public void removeItem(String name)
    {
        items.removeValue(Items.get(name), true);
    }

    public static final class Item 
    {
        private Inventory inventory = null;
        @NotNull
        public final String name;
        @NotNull
        public final String icon;
        @Nullable
        public final String description;
        @Nullable
        public final ItemListener listener;

        public Item()
        {
            this("INVALID", "invalid_item", "Invalid Item");
        }

        protected Inventory Inventory() 
        {
            return inventory;
        }

        public Item(@NotNull String name, @NotNull String icon, @Nullable String description, @Nullable ItemListener listener)
        {
            this.name = name;
            this.icon = icon;
            this.description = description;
            this.listener = listener;
        }

        public Item(@NotNull String name, @NotNull String icon, @Nullable String description)
        {
            this(name, icon, description, null);
        }
    }

    public interface ItemListener
    {
        void itemClicked();
    }
}

File: core/src/de/project/ice/inventory/Inventory.kt
package de.project.ice.inventory

import com.badlogic.gdx.utils.Array
import de.project.ice.IceGame

class Inventory(private val game: IceGame) {
    val items = Array<Item>()

    fun addItem(name: String) {
        removeItem(name)
        val item = Items.get(name)
        item.inventory = this
        items.add(item)
    }

    fun removeItem(name: String) {
        items.removeValue(Items.get(name), true)
    }

    class Item @JvmOverloads constructor(
            val name: String = "INVALID",
            val icon: String = "invalid_item",
            val description: String = "Invalid Item",
            val listener: ItemListener? = null) {

        internal var inventory: Inventory? = null

        protected fun Inventory(): Inventory? {
            return inventory
        }

        override fun equals(other: Any?): Boolean{
            if (this === other) return true
            if (other?.javaClass != javaClass) return false

            other as Item

            if (name != other.name) return false

            return true
        }

        override fun hashCode(): Int{
            return name.hashCode()
        }


    }

    interface ItemListener {
        fun itemClicked(game: IceGame)
    }
}
